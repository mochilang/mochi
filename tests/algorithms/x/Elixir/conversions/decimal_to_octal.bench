{
  "duration_us": 7119,
  "memory_bytes": 25576,
  "name": "main"
}