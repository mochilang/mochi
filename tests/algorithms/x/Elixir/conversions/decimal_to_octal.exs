# Code generated by Mochi transpiler 2025-08-06 23:01 +0700
defmodule Main do
  def int_pow(base, exp) do
    try do
      result = 1
      i = 0
      while_fun = fn while_fun, i, result ->
        if i < exp do
          result = result * base
          i = i + 1
          while_fun.(while_fun, i, result)
        else
          {i, result}
        end
      end
      {i, result} = try do
          while_fun.(while_fun, i, result)
        catch
          {:break, {i, result}} -> {i, result}
        end

      throw {:return, result}
    catch
      {:return, val} -> val
    end
  end
  def decimal_to_octal(num) do
    try do
      if num == 0 do
        throw {:return, "0o0"}
      end
      octal = 0
      counter = 0
      value = num
      while_fun_2 = fn while_fun_2, counter, octal, value ->
        if value > 0 do
          remainder = rem(value, 8)
          octal = octal + remainder * int_pow(10, counter)
          counter = counter + 1
          value = div(value, 8)
          while_fun_2.(while_fun_2, counter, octal, value)
        else
          {counter, octal, value}
        end
      end
      {counter, octal, value} = try do
          while_fun_2.(while_fun_2, counter, octal, value)
        catch
          {:break, {counter, octal, value}} -> {counter, octal, value}
        end

      throw {:return, ("0o" <> Kernel.to_string(octal))}
    catch
      {:return, val} -> val
    end
  end
  def main() do
    IO.puts(Kernel.inspect(decimal_to_octal(2)))
    IO.puts(Kernel.inspect(decimal_to_octal(8)))
    IO.puts(Kernel.inspect(decimal_to_octal(65)))
    IO.puts(Kernel.inspect(decimal_to_octal(216)))
    IO.puts(Kernel.inspect(decimal_to_octal(512)))
  end
end
Main.main()
