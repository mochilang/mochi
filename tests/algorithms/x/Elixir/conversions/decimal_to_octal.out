warning: variable "i" is unused (there is a variable with the same name in the context, use the pin operator (^) to match on it or prefix this variable with underscore if it is not meant to be used)
    │
 16 │       {i, result} = try do
    │        ~
    │
    └─ /workspace/mochi/tests/algorithms/x/Elixir/conversions/decimal_to_octal.exs:16:8: Main.int_pow/2

    warning: variable "counter" is unused (there is a variable with the same name in the context, use the pin operator (^) to match on it or prefix this variable with underscore if it is not meant to be used)
    │
 46 │       {counter, octal, value} = try do
    │        ~
    │
    └─ /workspace/mochi/tests/algorithms/x/Elixir/conversions/decimal_to_octal.exs:46:8: Main.decimal_to_octal/1

    warning: variable "value" is unused (there is a variable with the same name in the context, use the pin operator (^) to match on it or prefix this variable with underscore if it is not meant to be used)
    │
 46 │       {counter, octal, value} = try do
    │                        ~
    │
    └─ /workspace/mochi/tests/algorithms/x/Elixir/conversions/decimal_to_octal.exs:46:24: Main.decimal_to_octal/1

"0o2"
"0o10"
"0o101"
"0o330"
"0o1000"